File: modules/printer-drivers/android/src/main/java/expo/modules/printerdrivers/utils/helpers/WoosimHelper.kt
package expo.modules.printerdrivers.utils.helpers

import android.content.Context
import android.graphics.BitmapFactory
import com.woosim.printer.WoosimCmd
import com.woosim.printer.WoosimImage
import java.io.File
import java.io.FileInputStream
import java.nio.ByteBuffer
import java.nio.charset.Charset
import java.nio.charset.StandardCharsets

object WoosimHelper {
    fun addAlignedString(
        str: String = "",
        align: Int = WoosimCmd.ALIGN_LEFT,
        bold: Boolean = false,
        doubleHeight: Boolean = false,
        charset: Charset = StandardCharsets.UTF_8,
    ): ByteArray {
        val estimatedSize = 64 + str.toByteArray(charset).size
        val buffer = ByteBuffer.allocate(estimatedSize)

        // Đặt font nhỏ nhất trước khi in
        buffer.put(WoosimCmd.extendFont(1, 1))

        fun setTextAttributes(bold: Boolean, doubleHeight: Boolean, align: Int) {
            if (bold) buffer.put(WoosimCmd.setBold(true))
            if (doubleHeight) buffer.put(WoosimCmd.extendFont(1, 2))
            buffer.put(WoosimCmd.setTextAlign(align))
        }

        fun resetTextAttributes(bold: Boolean, doubleHeight: Boolean) {
            if (bold) buffer.put(WoosimCmd.setBold(false))
            if (doubleHeight) buffer.put(WoosimCmd.extendFont(1, 1))
            buffer.put(WoosimCmd.setTextAlign(WoosimCmd.ALIGN_LEFT))
        }

        setTextAttributes(bold, doubleHeight, align)
        buffer.put(str.toByteArray(charset))
        resetTextAttributes(bold, doubleHeight)

        return CommonHelper.convertBufferToByteArray(buffer)
    }

    fun addTwoMarginedStrings(
        str1: String = "",
        str2: String = "",
        bold: Boolean = false,
        doubleHeight: Boolean = false,
        len: Int,
        charset: Charset = StandardCharsets.UTF_8,
    ): ByteArray {
        val estimatedSize = 64 + str1.toByteArray(charset).size + str2.toByteArray(charset).size
        val buffer = ByteBuffer.allocate(estimatedSize)

        fun setTextAttributes(bold: Boolean, doubleHeight: Boolean) {
            if (bold) buffer.put(WoosimCmd.setBold(true))
            if (doubleHeight) buffer.put(WoosimCmd.extendFont(1, 2))
        }

        fun resetTextAttributes(bold: Boolean, doubleHeight: Boolean) {
            if (bold) buffer.put(WoosimCmd.setBold(false))
            if (doubleHeight) buffer.put(WoosimCmd.extendFont(1, 1))
        }

        setTextAttributes(bold, doubleHeight)

        // Calculate actual lengths
        val str1Length = str1.length
        val str2Length = str2.length

        if (str1Length + str2Length >= len) {
            // If combined length exceeds limit, try to fit by removing space
            if (str1Length + str2Length == len + 1) {
                // Case where removing one space will make it fit
                buffer.put(WoosimCmd.setTextAlign(WoosimCmd.ALIGN_LEFT))
                buffer.put(str1.ifBlank { " " }.toByteArray(charset))
                buffer.put(WoosimCmd.setTextAlign(WoosimCmd.ALIGN_RIGHT))
                buffer.put(str2.ifBlank { " " }.toByteArray(charset))
                buffer.put(WoosimCmd.setTextAlign(WoosimCmd.ALIGN_LEFT))
            } else {
                // If still too long even without space, then wrap
                val str = CommonHelper.createWrappedString("$str1 $str2", len)
                buffer.put(WoosimCmd.setTextAlign(WoosimCmd.ALIGN_LEFT))
                buffer.put(str.toByteArray(charset))
                buffer.put(WoosimCmd.printLineFeed(0))
            }
        } else {
            // Normal case - enough space for both strings
            buffer.put(WoosimCmd.setTextAlign(WoosimCmd.ALIGN_LEFT))
            buffer.put(str1.ifBlank { " " }.toByteArray(charset))
            buffer.put(WoosimCmd.setTextAlign(WoosimCmd.ALIGN_RIGHT))
            buffer.put(str2.ifBlank { " " }.toByteArray(charset))
            buffer.put(WoosimCmd.setTextAlign(WoosimCmd.ALIGN_LEFT))
        }

        resetTextAttributes(bold, doubleHeight)

        return CommonHelper.convertBufferToByteArray(buffer)
    }

    fun addImage(context: Context, fileName: String): ByteArray {
        val imageFilePath = "${context.cacheDir}/$fileName"
        val imageFile = File(imageFilePath)

        val bitmap = FileInputStream(imageFile).use { inputStream ->
            BitmapFactory.decodeStream(inputStream)
        }

        val arrBitmap = WoosimImage.printStdModeBitmap(bitmap)
        bitmap.recycle()

        val buffer = ByteBuffer.allocate(64 + arrBitmap.size).apply {
            put("\n".toByteArray()) // Have to add this to make sure the image is really center
            put(WoosimCmd.setAlignment(WoosimCmd.ALIGN_CENTER))
            put(arrBitmap)
            put(WoosimCmd.setAlignment(WoosimCmd.ALIGN_LEFT))
        }

        return CommonHelper.convertBufferToByteArray(buffer)
    }

    fun addThreeMarginedStrings(
        leftString: String,
        middleString: String,
        rightString: String,
        len: Int,
        leftBold: Boolean = false,
        middleBold: Boolean = false,
        rightBold: Boolean = false,
        allBold: Boolean = false,
        charset: Charset = StandardCharsets.UTF_8
    ): ByteArray {
        val left = leftString.trim { it <= ' ' }
        val middle = middleString.trim { it <= ' ' }
        val right = rightString.trim { it <= ' ' }

        val estimatedSize = 64 + left.toByteArray(charset).size + middle.toByteArray(charset).size + right.toByteArray(charset).size
        val buffer = ByteBuffer.allocate(estimatedSize)

        try {
            val totalLen = left.length + middle.length + right.length
            if (totalLen > len) {
                // Overflow: wrap the combined text
                val wrapped = CommonHelper.createWrappedString("$left $middle $right", len)
                buffer.put(WoosimCmd.setTextAlign(WoosimCmd.ALIGN_LEFT))
                buffer.put(wrapped.toByteArray(charset))
                buffer.put(WoosimCmd.printLineFeed(0))
            } else {
                // Calculate spaces
                val sp = len / 2 + middle.length / 2 - (left.length + middle.length)
                val sp1 = maxOf(1, sp)
                val tempRe = (left + " ".repeat(sp1) + middle).trim { it <= ' ' }
                val sp2 = maxOf(1, len - (tempRe.length + right.length))

                val space1 = " ".repeat(sp1)
                val space2 = " ".repeat(sp2)

                buffer.put(WoosimCmd.setTextAlign(WoosimCmd.ALIGN_LEFT))

                // Left
                if (allBold || leftBold) buffer.put(WoosimCmd.setBold(true))
                buffer.put(left.ifBlank { " " }.toByteArray(charset))
                if (allBold || leftBold) buffer.put(WoosimCmd.setBold(false))

                // Space between left and middle
                buffer.put(space1.toByteArray(charset))

                // Middle
                if (allBold || middleBold) buffer.put(WoosimCmd.setBold(true))
                buffer.put(middle.ifBlank { " " }.toByteArray(charset))
                if (allBold || middleBold) buffer.put(WoosimCmd.setBold(false))

                // Space between middle and right
                buffer.put(space2.toByteArray(charset))

                // Right
                if (allBold || rightBold) buffer.put(WoosimCmd.setBold(true))
                buffer.put(right.ifBlank { " " }.toByteArray(charset))
                if (allBold || rightBold) buffer.put(WoosimCmd.setBold(false))

                buffer.put(WoosimCmd.setTextAlign(WoosimCmd.ALIGN_LEFT))
            }
        } catch (ex: Exception) {
            ex.printStackTrace()
        }

        return CommonHelper.convertBufferToByteArray(buffer)
    }

    fun addLineFeed(lines: Int = 1): ByteArray {
        return WoosimCmd.printLineFeed(lines)
    }
}

File: modules/printer-drivers/ios/helpers/WoosimHelper.swift
import Foundation
import UIKit
import woosim302

enum WoosimHelper {
    static func addAlignedString(
        _ str: String = "",
        align: WSCmd.AlignType = .Left,
        bold: Bool = false,
        doubleHeight: Bool = false
    ) -> Data {
        var out = Data()

        out.append(WSCmd.extendFont(width: 1, height: 1))

        if bold { out.append(WSCmd.setBold(true)) }
        if doubleHeight { out.append(WSCmd.extendFont(width: 1, height: 2)) }
        out.append(WSCmd.alignText(align))

        if let bytes = str.data(using: .utf8) {
            out.append(bytes)
        }

        if bold { out.append(WSCmd.setBold(false)) }
        if doubleHeight { out.append(WSCmd.extendFont(width: 1, height: 1)) }
        out.append(WSCmd.alignText(.Left))

        return out
    }

    static func addTwoMarginedStrings(
        _ str1: String = "",
        _ str2: String = "",
        bold: Bool = false,
        doubleHeight: Bool = false,
        len: Int
    ) -> Data {
        var out = Data()

        if bold { out.append(WSCmd.setBold(true)) }
        if doubleHeight { out.append(WSCmd.extendFont(width: 1, height: 2)) }

        let l1 = str1.count
        let l2 = str2.count

        if l1 + l2 >= len {
            if l1 + l2 == len + 1 {
                out.append(WSCmd.alignText(.Left))
                out.append((str1.isEmpty ? " " : str1).data(using: .utf8) ?? Data())
                out.append(WSCmd.alignText(.Right))
                out.append((str2.isEmpty ? " " : str2).data(using: .utf8) ?? Data())
                out.append(WSCmd.alignText(.Left))
            } else {
                let wrapped = CommonHelper.createWrappedString("\(str1) \(str2)", len: len)
                out.append(WSCmd.alignText(.Left))
                out.append(wrapped.data(using: .utf8) ?? Data())
                out.append(WSCmd.printFeed(line: 0))
            }
        } else {
            out.append(WSCmd.alignText(.Left))
            out.append((str1.isEmpty ? " " : str1).data(using: .utf8) ?? Data())
            out.append(WSCmd.alignText(.Right))
            out.append((str2.isEmpty ? " " : str2).data(using: .utf8) ?? Data())
            out.append(WSCmd.alignText(.Left))
        }

        if bold { out.append(WSCmd.setBold(false)) }
        if doubleHeight { out.append(WSCmd.extendFont(width: 1, height: 1)) }

        return out
    }

    static func addThreeMarginedStrings(
        _ leftString: String,
        _ middleString: String,
        _ rightString: String,
        pageLength: Int,
        leftBold: Bool = false,
        middleBold: Bool = false,
        rightBold: Bool = false,
        allBold: Bool = false
    ) -> Data {
        let left = leftString.trimmingCharacters(in: .whitespaces)
        let middle = middleString.trimmingCharacters(in: .whitespaces)
        let right = rightString.trimmingCharacters(in: .whitespaces)

        var out = Data()
        let total = left.count + middle.count + right.count

        if total > pageLength {
            let wrapped = CommonHelper.createWrappedString(
                "\(left) \(middle) \(right)", len: pageLength
            )
            out.append(WSCmd.alignText(.Left))
            out.append(wrapped.data(using: .utf8) ?? Data())
            out.append(WSCmd.printFeed(line: 0))
            return out
        }

        let middleStart = max(pageLength / 2 - middle.count / 2, 0)
        let sp1 = max(middleStart - left.count, 1)
        let middleEnd = middleStart + middle.count
        let rightStart = pageLength - right.count
        let sp2 = max(rightStart - middleEnd, 1)

        let space1 = String(repeating: " ", count: sp1)
        let space2 = String(repeating: " ", count: sp2)

        out.append(WSCmd.alignText(.Left))

        if allBold || leftBold { out.append(WSCmd.setBold(true)) }
        out.append((left.isEmpty ? " " : left).data(using: .utf8) ?? Data())
        if allBold || leftBold { out.append(WSCmd.setBold(false)) }

        out.append(space1.data(using: .utf8) ?? Data())

        if allBold || middleBold { out.append(WSCmd.setBold(true)) }
        out.append((middle.isEmpty ? " " : middle).data(using: .utf8) ?? Data())
        if allBold || middleBold { out.append(WSCmd.setBold(false)) }

        out.append(space2.data(using: .utf8) ?? Data())

        if allBold || rightBold { out.append(WSCmd.setBold(true)) }
        let rightWithNL = right.isEmpty ? " " : "\(right)\n"
        out.append(rightWithNL.data(using: .utf8) ?? Data())
        if allBold || rightBold { out.append(WSCmd.setBold(false)) }

        out.append(WSCmd.alignText(.Left))
        return out
    }

    static func addImage(fileName: String, align: WSCmd.AlignType = .Center) -> Data {
        var out = Data()

        guard let cachesDir = NSSearchPathForDirectoriesInDomains(
            .cachesDirectory, .userDomainMask, true
        ).first else {
            NSLog("[WoosimHelper] --> caches dir unavailable")
            return out
        }

        let imagePath = (cachesDir as NSString).appendingPathComponent(fileName)
        guard let image = UIImage(contentsOfFile: imagePath) else {
            NSLog("[WoosimHelper] --> failed to load image at \(imagePath)")
            return out
        }
        
        out.append("\n".data(using: .utf8) ?? Data())
        out.append(WSCmd.alignContent(align))
        out.append(WSImage.printImage(image, dithering: true, compress: true))
        out.append(WSCmd.alignContent(.Left))

        return out
    }
}
